### ISSUE TITLE
Watches don't seem to detect diffs made outside of Terraform

### ISSUE THREAD

**Issue by tdmalone** [phillbaker/terraform-provider-elasticsearch#65]
Thanks again for putting this provider together! I'm finally getting around to implementing it into our workflows and am looking forward to no longer having to version our Elastic resources 'manually' :)

Doing some quick testing, and I think I'm unable to get the provider to generate a diff on a watch that has been edited outside of Terraform. This is obviously a key part of managing the resources via Terraform, that we want to be able to detected unintended drift in the live resources.

Firstly, some necessary data:
- Terraform v0.11.14
- latest release of terraform-provider-elasticsearch (I think - installed it fresh today, unsure how to 100% check the version though as `terraform version` shows 'unversioned' for the provider)
- cluster is on Elastic Cloud
- `sniff` is set to `false` in provider config
- cluster username and password are provided via environment variables
- there are no other provider settings configured

What I did:
1. Successfully created a new watch via Terraform, with a very simple config (see below)
1. Adjusted in Terraform the value of `trigger.schedule.interval` to 100m, to confirm that the diff is detected and updated (works great!)
1. Adjusted _manually via Kibana_ the value of `trigger.schedule.interval` to 102m, and ran `terraform apply` again. At this point I would expect Terraform to want to set the interval back to 100m again - however it detects no diff; no changes to apply (at this point I also tried changing it back to 10m in my code, Terraform then generates a diff but it notes that I'd be changing '100m' to '10m' - i.e. it isn't aware of the 102m that I set manually in the meantime.

It _seems_ like a bug - I can't think of something I've wrong here as it's quite a simple test - but would appreciate any insight you can offer. I'm very inexperienced in Go but I might be able to offer a PR given a few pointers!

Watch config used:
```
resource "elasticsearch_watch" "test" {
  watch_id = "test-watch"

  body = <<EOF
    {
      "trigger": {
        "schedule": {
          "interval": "10m"
        }
      },
      "input": {
        "search": {
          "request": {
            "indices": ["filebeat-*"],
            "body": {
              "query": {
                "match_all" : {}
              }
            }
          }
        }
      },
      "condition": {
        "always": {}
      },
      "actions": {},
      "metadata": {
        "name": "test"
      }
    }
  EOF
}
```

**Comment by phillbaker** [comment]
Hm, this might be because terraform is storing the watch as the JSON for the entire resource, instead of breaking out the attributes and then on the plan, comparing each attribute. (Admittedly this is not the best way to do things - it's a bit of a hack to support things quickly.) However, I would expect there to be a diff in the entire JSON body attribute.

Can you try doing a [`terraform refresh`](https://www.terraform.io/docs/commands/refresh.html) in between (2) and (3)?

**Comment by tdmalone** [comment]
I've tried replicating this again with a `terraform refresh` before step 3 - same result.
```
$ terraform refresh
elasticsearch_watch.test: Refreshing state... (ID: test-watch)

$ terraform apply
elasticsearch_watch.test: Refreshing state... (ID: test-watch)

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
Releasing state lock. This may take a few moments...
```
(it looks like an apply is doing the refresh anyway)

Supporting this quickly is better than not supporting things at all! We can work with this, we just need to be careful not to modify anything directly in Elastic.

Ideally we would still be able to detect drift though, I'm happy to try to work on something for this but just not sure where to start at the moment. I might have a look through the source and see if I can figure things out. I would have thought also, though, that a diff would still be shown on the entire JSON.

**Comment by phillbaker** [comment]
@tdmalone can you confirm the version of elasticsearch/kibana this was seen on?

**Comment by phillbaker** [comment]
I actually think may be related to try to set a non-primative type in the state:
https://github.com/phillbaker/terraform-provider-elasticsearch/blob/dc29f7b9f2b3b5ca64a60fa6a037a5d2bdb0e616/es/resource_elasticsearch_xpack_watch.go#L100

(we're not checking the err here either, which we likely do have)

**Comment by tdmalone** [comment]
> @tdmalone can you confirm the version of elasticsearch/kibana this was seen on?

Apologies for the delay - this is on Elasticsearch 7.6.2.
If you're unable to replicate I could try upgrading to 7.7.1?

**Comment by phillbaker** [comment]
I have a fix in progress for this, but it's blocked by an upstream bug, PR to fix that issue is here: https://github.com/olivere/elastic/pull/1345.

That fix will also need to be ported to the v6 version of the upstream library.

**Comment by tjbaker** [comment]
I noticed the same problem with monitors, refresh/plan is not detecting drift here either.
I just upgraded to 1.3 and will see if that makes a difference.

**Comment by phillbaker** [comment]
Yes, unfortunately, monitors is due to a similar issue here:
https://github.com/phillbaker/terraform-provider-elasticsearch/blob/1e7b35a99451a84f9620a316d3fa30fe36b7d4e9/es/resource_elasticsearch_opendistro_monitor.go#L81 

These go objects cannot be serialized to terraform state, they must be `map[string]interface{}` objects. A branch to address several of these is here: https://github.com/phillbaker/terraform-provider-elasticsearch/tree/check-set-errs

**Comment by phillbaker** [comment]
It looks like one complication with correctly comparing monitors is that opendistro saves the default values of search queries in the monitor definition. So for example, `adjust_pure_negative`, even if not specified by the plan, comes back in the response.

**Comment by tjbaker** [comment]
Regarding monitors, something else that bit me when in upgrading to 1.03 and the new resource names per the deprecation warnings....monitors have an id reference to the destination embedded in their body.  When you change the provider resource for destinations (they now have opendistro in the name) it destroys and recreates the destination so it gets a new `id`.  But the monitor doesn't notice that the destination changed...so it now has a reference to a non-existent destination and is broken.

**Comment by phillbaker** [comment]
Upstream issue for xpack watch was fixed for v7 in https://github.com/olivere/elastic/pull/1363, v6 fix is in https://github.com/olivere/elastic/pull/1368.

**Comment by phillbaker** [comment]
>  When you change the provider resource for destinations (they now have opendistro in the name) it destroys and recreates the destination so it gets a new id

@tjbaker would just not destroying and recreating, but just `terraform state mv` the resources.

**Comment by phillbaker** [comment]
Released a fix for the watch diff in 1.4.2, note that it looks like the watch API may return default values that were not passed in the original request, e.g. for log actions, `"level": "info"`, which would result in a perpetual diff unless it's pulled into the definition.

Still working on the monitor issue.

**Comment by tdmalone** [comment]
@phillbaker Thanks - it works great! I am indeed seeing those diffs - with eg. `rest_total_hits_as_int` and `search_type`, which have default values if they're not set. I think it's possible for Terraform to store the returned result in the state, after an apply, so that it doesn't cause a perpetual diff. Unfortunately I'm not sure how to do it... I just know that I've seen it in discussions around other providers.

**Comment by tdmalone** [comment]
Actually - I am seeing _quite_ a lot of diffs - too many to adjust in our configurations, so I might need to downgrade to the previous version for now. Other examples include setting `"lang": "painless"` on transform scripts (the default value if it's not specified), as well as unsetting most of the computed values for `webhook` actions where we've just provided a `url` value.

**Comment by phillbaker** [comment]
Part of the issue is that, originally, in order to add support for this functionality quickly, the resource was built to accept just a JSON blob, instead of breaking out the attributes as most resources do. The downside of this is that it's harder to use terraform schema helpers to deal with issues like this. The medium term fix is to transition away from the json blob to a "real" schema.

**Comment by phillbaker** [comment]
Closing as this issue is fixed, further improvements will be in  #93
